### PR TITLE
1098 shell to app

### LINKED ISSUE #1098: Shell into application pod
**Is your feature request related to a problem? Please describe.**
I would like to be able to get shell access into a running application instance without needing to figure out my pod name and running `kubectl exec`.

This would be helpful for debugging issues where having access to the environment makes the most sense (networking, unexpected or broken builds, etc...).

**Describe the solution you'd like**
I would like to be able to run `epinio app shell <app_name> [-c <index>]` and get a terminal prompt. I would also like to add this functionality to the VS Code plugin without requiring the epinio-cli to be in the path. 

**Describe alternatives you've considered**
I thought about having the command just print the correct kubectl command to run but I believe this doesn't give a good solution. It's certainly workable but looks "cheap". We also don't want to have kubectl be a requirement for end users.

### PR DESCRIPTION
**Pull request by jimmykarily**
Fixes #1098 

Doesn't handle multi-instance apps. For now, it simply picks one of the application pods. Multi-instances needs additional commands to be implemented, so that the user can list the instances (pods?) that are available.

It can be implemented as an optional argument to `epinio apps exec`. When set it will be the desired pod's name. If not set, we will pick one randomly. If agreed, a new issue will be created for that.

### CODE REVIEW

File: internal/api/v1/application/exec.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		Resource("pods").
+		Name(podName).
+		SubResource("exec").
+		VersionedParams(&v1.PodExecOptions{
+			Stdin:     true,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       true,
+			Container: container,
+			// TODO: What if the container doesn't have bash?
```
Comment (jimmykarily) [inline internal/api/v1/application/exec.go:83]: I guess this depends on the builder image and what kind of base image is used for the applications. Maybe we should leave this as it is until this is a problem (e.g. someone wants to use a custom builder image that produces images without bash?)

File: Makefile [hunk 2]
```diff
@@ -99,7 +99,7 @@ test-acceptance: showfocus
 	ginkgo -nodes ${GINKGO_NODES} -stream -slowSpecThreshold ${GINKGO_SLOW_TRESHOLD} -randomizeAllSpecs --flakeAttempts=${FLAKE_ATTEMPTS} -failOnPending acceptance/. acceptance/api/v1/. acceptance/apps/.
 
 test-acceptance-api: showfocus
-	ginkgo -nodes ${GINKGO_NODES} -stream -slowSpecThreshold ${GINKGO_SLOW_TRESHOLD} -randomizeAllSpecs --flakeAttempts=${FLAKE_ATTEMPTS} -failOnPending acceptance/api/v1/.
+	ginkgo -v -nodes ${GINKGO_NODES} -stream -slowSpecThreshold ${GINKGO_SLOW_TRESHOLD} -randomizeAllSpecs --flakeAttempts=${FLAKE_ATTEMPTS} -failOnPending acceptance/api/v1/.
```
Comment (enrichman) [inline Makefile:102]: Left verbose?

File: internal/cli/apps.go [hunk 3]
```diff
@@ -173,6 +174,34 @@ var CmdAppLogs = &cobra.Command{
...
+		client, err := usercmd.New()
+		if err != nil {
+			return errors.Wrap(err, "error initializing cli")
+		}
+
+		// TODO: "--instance" ? to define the app instance (the pod)
+		// follow, err := cmd.Flags().GetBool("follow")
+		// if err != nil {
+		// 	return errors.Wrap(err, "error reading option --follow")
+		// }
```
Comment (enrichman) [inline internal/cli/apps.go:194]: Can we remove this comment?

File: internal/cli/apps.go [hunk 4]
```diff
@@ -173,6 +174,34 @@ var CmdAppLogs = &cobra.Command{
...
+
+		err = client.AppExec(cmd.Context(), args[0])
+		if err != nil {
+			return errors.Wrap(err, "error getting a shell to application")
```
Comment (enrichman) [inline internal/cli/apps.go:198]: Since errors.Wrap return `nil` in case of a nil error we can avoid the check (see the above function).

**Comment by jimmykarily** [comment]
I'm trying to fix the broken test. For some reason, we get CR characters in random places when we read the remote stdout. I'm trying to understand why.

File: Makefile (hunk 2 above)
Comment (jimmykarily) [reply Makefile:102]: Yes, fixing.

File: internal/cli/apps.go (hunk 3 above)
Comment (jimmykarily) [reply internal/cli/apps.go:194]: Extracted to a new issue: https://github.com/epinio/epinio/issues/1151

removing TODO

File: internal/cli/apps.go (hunk 4 above)
Comment (jimmykarily) [reply internal/cli/apps.go:198]: oh nice, fixing.

File: internal/api/v1/application/exec.go (hunk 1 above)
Comment (jimmykarily) [reply internal/api/v1/application/exec.go]: Removing `TODO`. We will fix this if it's ever a problem

**Review by enrichman — APPROVED** [review]

**Pull request merged**